File: machine-learning/ex1/python/gradientDescent.py
import numpy as np
from computeCost import computeCost

def gradientDescent(X, y, theta=np.zeros(2), alpha=0.01, iterations=10000):
    m = len(y)
    J_history = np.zeros(iterations)

    for itr in np.arange(iterations):
        hypothesis = X.dot(theta)
        error = hypothesis - np.transpose([y])

        gradient = (1.0/m) * X.T.dot(error)
        theta = theta - alpha * gradient
        
        J_history[itr] = computeCost(X, y, theta)

    return theta, J_history

File: machine-learning/ex1/gradientDescent.m
function [theta, J_history] = gradientDescent(X, y, theta, alpha, num_iters)

    m = length(y); % number of training examples
    J_history = zeros(num_iters, 1); % where we store our gradient answers

    for iter = 1:num_iters
      
        hypothesis = X * theta;
        error = hypothesis - y;
        % error = ((X * theta) - y);
        
        gradient = (1/m) .* X' * error;
        % gradient = (1/m).* X' * ((X * theta) - y);
        
        theta = theta - alpha .* gradient;

        % Save the cost J in every iteration    
        J_history(iter) = computeCost(X, y, theta);

    end
end
